zerver/lib/send_email.py

```diff
@@ -418,21 +418,40 @@ def send_custom_email(users: List[UserProfile], options: Dict[str, Any]) -> None
 
     inline_template(email_filename)
 
-    # Finally, we send the actual emails.
-    for user_profile in users:
-        if options.get("admins_only") and not user_profile.is_realm_admin:
-            continue
+    # Finally, we send the actual emails if its not a dry-run
+    if not options["dry_run"]:
+        for user_profile in users:
+            if options.get("admins_only") and not user_profile.is_realm_admin:
+                continue
+            context = {
+                "realm_uri": user_profile.realm.uri,
+                "realm_name": user_profile.realm.name,
+            }
+            send_email(
+                email_id,
+                to_user_ids=[user_profile.id],
+                from_address=FromAddress.SUPPORT,
+                reply_to_email=options.get("reply_to"),
+                from_name=get_header(
+                    options.get("from_name"), parsed_email_template.get("from"), "from_name"
+                ),
+                context=context,
+            )
+    else:
+        print("The email will be sent to the following users:")
+        for user_profile in users:
+            print(user_profile.email)
+
+        # get the plain/text part of the mail.
         context = {
-            "realm_uri": user_profile.realm.uri,
-            "realm_name": user_profile.realm.name,
+            "realm_uri": users[0].realm.uri,
+            "realm_name": users[0].realm.name,
         }
-        send_email(
+        mail = build_email(
             email_id,
-            to_user_ids=[user_profile.id],
-            from_address=FromAddress.SUPPORT,
-            reply_to_email=options.get("reply_to"),
-            from_name=get_header(
-                options.get("from_name"), parsed_email_template.get("from"), "from_name"
-            ),
-            context=context,
+            to_user_ids=[users[0].id],
+            context={},
         )
+
+        print("The email that will be sent is:")
+        print(mail.message().get_payload()[0])
```

Comment:
This structure ends up forking a lot of the code, in a way that could lead to bugs.  An alternative plan would, for example, to replace `send_email` with `build_email` in the original tool, and then just have the last line be either to send the build email or print its details depending on the dry_run state.  Can you try that approach out and see if you like it?

Reply:
Thanks a lot for the review @timabbott 
Now, I have made dry_run as an optional argument whose default value is `False` in the `send_email` function, it already calls `build_email`, depending on the status on `dry_run`, either the mail is sent or printed on the terminal